docs/madr/003-arns-undername-limits.md

```diff
@@ -0,0 +1,237 @@
+# ArNS Undername Limit Enforcement
+
+- Status: proposed
+- Deciders: [Ariel], [Dylan], [Atticus], [Phil], [David]
+- Date: 2025-01-14
+- Authors: [Dylan]
+
+## Context and Problem Statement
+
+ArNS names have a supported undername limit, defined by the ARIO network contract. Increasing this limit requires payment in $ARIO tokens to compensate gateway operators for the additional computational resources needed to serve undername and promote responsible usage of ArNS. AR.IO gateways must enforce this limit when resolving ArNS names to ensure operators are fairly rewarded for their services through fees paid by increasing a name's undername limit and providing a consistent experience across the network.
+
+Currently, the `getRecords` API on ANTs return **a table of records** allowing for efficient undername lookups. Lua does not guarantee the order of keys in a table, making it difficult to know what undernames to enforce. Sorting needs to either 1. be done by the ANT and provided as an array of records, or 2. be done by the [ar-io-node] using additional information provided by the ANT. Additionally, caching within the ar-io-node needs to properly handle these priority order changes.
```

Comment:
`API on ANTs return` -> `API on ANTs returns`

Reply:
this would requrie the ANT to check the contract for the limit, which requires a coroutine

Reply:
[90be780](https://github.com/ar-io/ar-io-node/pull/272/commits/90be780969e021f3f9cfee168c331a143129ec1e)